File: spark/src/main/scala/geotrellis/spark/io/file/FileValueReader.scala
package geotrellis.spark.io.file

import geotrellis.spark._
import geotrellis.spark.io._
import geotrellis.spark.io.avro.codecs.KeyValueRecordCodec
import geotrellis.spark.io.index._
import geotrellis.spark.io.avro._
import geotrellis.util.Filesystem

import org.apache.avro.Schema
import spray.json._
import spray.json.DefaultJsonProtocol._

import java.io.File
import scala.reflect.ClassTag

class FileValueReader(
  val attributeStore: AttributeStore,
  catalogPath: String
) extends ValueReader[LayerId] {

  def reader[K: AvroRecordCodec: JsonFormat: ClassTag, V: AvroRecordCodec](layerId: LayerId): Reader[K, V] = new Reader[K, V] {

    val header = attributeStore.readHeader[FileLayerHeader](layerId)
    val keyIndex = attributeStore.readKeyIndex[K](layerId)
    val writerSchema = attributeStore.readSchema(layerId)

    val maxWidth = Index.digits(keyIndex.toIndex(keyIndex.keyBounds.maxKey))
    val keyPath = KeyPathGenerator(catalogPath, header.path, keyIndex, maxWidth)

    def read(key: K): V = {
      val path = keyPath(key)

      if(!new File(path).exists)
        throw new TileNotFoundError(key, layerId)

      val bytes = Filesystem.slurp(path)
      val recs = AvroEncoder.fromBinary(writerSchema, bytes)(KeyValueRecordCodec[K, V])

      recs
        .find { case (recordKey, _) => recordKey == key }
        .map { case (_, recordValue) => recordValue }
        .getOrElse(throw new TileNotFoundError(key, layerId))
    }
  }
}

object FileValueReader {
  def apply[K: AvroRecordCodec: JsonFormat: ClassTag, V: AvroRecordCodec](
    attributeStore: AttributeStore,
    catalogPath: String,
    layerId: LayerId
  ): Reader[K, V] =
    new FileValueReader(attributeStore, catalogPath).reader(layerId)

  def apply(catalogPath: String): FileValueReader =
    new FileValueReader(new FileAttributeStore(catalogPath), catalogPath)

  def apply(attributeStore: FileAttributeStore): FileValueReader =
    new FileValueReader(attributeStore, attributeStore.catalogPath)
}

File: geotrellis-python/geotrellis/spark/io/file/FileValueReader.py
from geotrellis.spark.io.index.Index import Index
from geotrellis.spark.io.package_scala import TileNotFoundError
from geotrellis.spark.io.file.FileAttributeStore import FileAttributeStore
from geotrellis.spark.io.file.FileLayerHeader import FileLayerHeader
from geotrellis.python.util.utils import file_exists, find, to_pairs, from_file
from KeyPathGenerator import generate_key_path_func

def file_value_reader(first, second = None):
    def get_params():
        if second is None:
            if isinstance(first, FileAttributeStore):
                return first, first.catalogPath
            elif isinstance(first, str):
                return FileAttributeStore(first), first
            else:
                raise Exception("wrong params ({0}, {1})".format(str(first), str(second)))
        else:
            return first, second
    attribute_store, catalog_path = get_params()
    def reader(key_type, value_type, layer_id):
        header          = attribute_store.readHeader(FileLayerHeader, layer_id)
        key_index       = attribute_store.readKeyIndex(key_type, layer_id)
        writer_schema   = attribute_store.readSchema(layer_id)

        max_width = Index.digits(key_index.toIndex(key_index.keyBounds.maxKey))
        key_path = generate_key_path_func(catalog_path, header.path, key_index, max_width)
        def read_func(key):
            path = key_path(key)
            if not file_exists(path):
               raise TileNotFoundError(key, layer_id)
            recs = to_pairs(from_file(writer_schema, path))
            found = find(recs, lambda pair: pair._1 == key)
            if found is None:
                raise TileNotFoundError(key, layer_id)
            else:
                return found._2
        return Reader(read_func)
    return ReaderGenerator(reader)

class Reader(object):
    def __init__(self, read_func):
        self.read_func = read_func

    def read(self, key):
        return self.read_func(key)

class ReaderGenerator(object):
    def __init__(self, reader_func):
        self.reader_func = reader_func

    def reader(self, key_type, value_type, layer_id):
        return self.reader_func(key_type, value_type, layer_id)

